protovalidate/internal/constraints.py

```diff
@@ -269,24 +279,66 @@ def check_field_type(field: descriptor.FieldDescriptor, expected: int, wrapper_n
         raise CompilationError(msg)
 
 
+def _is_map(field: descriptor.FieldDescriptor):
+    return (
+        field.label == descriptor.FieldDescriptor.LABEL_REPEATED
+        and field.message_type is not None
+        and field.message_type.GetOptions().map_entry
+    )
+
+
+def _is_list(field: descriptor.FieldDescriptor):
+    return field.label == descriptor.FieldDescriptor.LABEL_REPEATED and not _is_map(field)
+
+
+def _zero_value(field: descriptor.FieldDescriptor, *, for_items: bool):
+    if for_items and _is_list(field):
+        message = message_factory.GetMessageClass(field.containing_type)()
+        setattr(message, field.name, [message_factory.GetMessageClass(field.message_type)()])
+        return message
+    elif field.message_type is not None and field.label != descriptor.FieldDescriptor.LABEL_REPEATED:
+        return message_factory.GetMessageClass(field.message_type)()
+    else:
+        return field.default_value
+
+
+def _proto_equals(a, b):
+    if isinstance(a, message.Message) and isinstance(b, message.Message):
+        # TODO: Python protobuf does not have the ability to compare messages.
```

Comment:
I just had a thought. While Python protobuf doesn't have a comparison function, celpy does. Maybe it would be better to convert the zero value to a celpy type: we are already calling `_field_value_to_cel` right after the zero check anyways. I'll try that next week.

Reply:
This works and is now committed, removing `_proto_equals` entirely.